src/why3_tactic.ml

```diff
@@ -0,0 +1,189 @@
+    | (_        , _       )                        ->
+        (* If the term [p] is mapped in [tbl] then use it. *)
+        try (tbl, Why3.Term.ps_app (List.assoc_eq Basics.eq t tbl) [])
+        with Not_found ->
+          (* Otherwise generate a new constant in why3. *)
+          let sym = Why3.Term.create_psymbol (Why3.Ident.id_fresh "P") [] in
+          ((t, sym)::tbl, Why3.Term.ps_app sym [])
+  in
+  match Basics.get_args t with
+  | (Symb(s,_), [t]) when s == cfg.symb_P -> Some (translate_prop tbl t)
+  | _                                     -> None
+
+(** [encode pos builtins hs g] translates the hypotheses [hs] and the goal [g]
+    into Why3 terms, to construct a Why3 task. *)
+let encode : Pos.popt -> sym StrMap.t -> Env.env -> term -> Why3.Task.task =
+    fun pos builtins hs g ->
+  let cfg = get_config pos builtins in
+  let (constants, hypothesis) =
+    let translate_hyp (tbl, map) (name, (_, hyp)) =
+      match translate_term cfg tbl (Bindlib.unbox hyp) with
+      | Some(tbl, why3_hyp) -> (tbl, StrMap.add name why3_hyp map)
+      | None                -> (tbl, map)
+    in
+    List.fold_left translate_hyp ([], StrMap.empty) hs
+  in
+  (* Translate the goal term. *)
+  let (tbl, why3_term) =
+    match translate_term cfg constants g with
+    | Some(tbl, why3_term) -> (tbl, why3_term)
+    | None                 ->
+        fatal pos "The goal [%a] is not of the form [P _]" Print.pp g
+  in
+  (* Add the declaration of every constant in a task. *)
+  let fn tsk (_,t) = Why3.Task.add_param_decl tsk t in
+  let tsk = List.fold_left fn None tbl in
+  (* Add the declaration of every hypothesis in the task. *)
+  let fn name t tsk =
+    let axiom = Why3.Decl.create_prsymbol (Why3.Ident.id_fresh name) in
+    Why3.Task.add_prop_decl tsk Why3.Decl.Paxiom axiom t
+  in
+  let tsk = StrMap.fold fn hypothesis tsk in
+  (* Add the goal itself. *)
+  let goal = Why3.Decl.create_prsymbol (Why3.Ident.id_fresh "main_goal") in
+  (* Return the task that contains the encoded lambdapi formula in Why3. *)
+  Why3.Task.add_prop_decl tsk Why3.Decl.Pgoal goal why3_term
+
+(** [run_task tsk pos prover_name] Run the task [tsk] with the specified
+    prover named [prover_name] and return the answer of the prover. *)
+let run_task : Why3.Task.task -> Pos.popt -> string -> bool =
+    fun tsk pos prover_name ->
+  (* Filter the set of why3 provers. *)
+  let filter = Why3.Whyconf.parse_filter_prover prover_name in
+  let provers = Why3.Whyconf.filter_provers why3_config filter in
+  (* Fail if we did not find a matching prover. *)
+  if Why3.Whyconf.Mprover.is_empty provers then
+    fatal pos "[%s] not installed or not configured" prover_name;
+  (* Return the prover configuration and load the driver. *)
+  let prover = snd (Why3.Whyconf.Mprover.max_binding provers) in
+  let driver =
+    try Why3.Whyconf.(load_driver why3_main why3_env prover.driver [])
+    with e -> fatal pos "Failed to load driver for %s (%a)"
```

Comment:
I made most of the last changes I wanted, but before we merge I would like you (@elhaddadyacine) to clarify what exceptions can be raised here. In general it is quite bad practice to have wildcards like these (for example, because of possible asynchronous exceptions).

Reply:
I did not find any specific exceptions that can be raised here since it was not mentioned in the Why3 documentation. I tried to raise some by changing/deleting driver files but it raised a fatal errors rather than an exception.

Reply:
OK, I'll have a look another time. I merged.